### ISSUE TITLE
[Bug] High-throughput architectures (IMPALA, APPO) regression

### ISSUE THREAD

**Issue by vakker** [ray-project/ray#22557]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.


### Ray Component

RLlib

### What happened + What you expected to happen

I was doing some experiments, and for some reason I couldn't get APPO to work properly, while I can get PPO to work quite reasonably. This led me to some debugging, and currently to me it seems that APPO might not be working at all, and potentially IMPALA is significantly less efficient than PPO (which might be a separate issue).

Looking at the reward curves (red - PPO, orange - IMPALA, pink - APPO), it seems that:

1. PPO trains the fastest in sample time (x axis - timesteps):
![rewards1](https://user-images.githubusercontent.com/3268882/155228257-6b0dd831-9304-4869-a44f-89c39b6a775b.png)

2. PPO vs IMPALA is almost the same in wall time (x axis - hours):
![rewards2](https://user-images.githubusercontent.com/3268882/155228256-5b47626b-8c0e-4184-ba35-a1595b8a2f8d.png)

3. The sample throughput seems to make sense for PPO and IMPALA, but it's disturbingly low for APPO:
![samples](https://user-images.githubusercontent.com/3268882/155228253-eb38ca5e-5d11-4101-a33e-0cf9778b0069.png)

Also, IMPALA is nowhere near the reported wall time ([here](https://github.com/ray-project/rl-experiments#pong-in-3-minutes)), while I'm using 32 workers. Well, I have 24 cores, and I set the `num_cpus_per_worker` to 0.5, because it doesn't saturate my system anyway, but I can run a benchmark on a 32 core machine if needed:
![cpu](https://user-images.githubusercontent.com/3268882/155229223-db754fde-0a90-4e6a-8527-58356b60c5f8.png)

Also, the GPU utilization seems a bit odd (I have 2 1080ti GPUs, but trained all 3 just on 1 GPU, I don't know why they show up on both graphs):
![gpu](https://user-images.githubusercontent.com/3268882/155229425-11a2dd4e-8da5-4788-881c-e08d801e7533.png)


So, I can partially explain the speed of IMPALA, probably my GPU cannot process the samples fast enough, so the overall training speed is lower than the reported. However, it shouldn't be lower then PPO, but I'm aware that they are different algorithms, so it's not completely apples to apples.

On the other hand, APPO doesn't seem to learn at all, which is odd. Also, to add to that, with APPO the training ends with OOM:
```
Using FIFO scheduling algorithm.                                                                                                                                                                          [52/9617]
Resources requested: 17.0/24 CPUs, 1.0/1 GPUs, 0.0/36.24 GiB heap, 0.0/18.12 GiB objects                                                                                                                           
Result logdir: /home/user/ray_results/pong-appo                                                                                                                                                                   
Number of trials: 1/1 (1 RUNNING)                                                                                                                                                                                  
+-------------------------------------+----------+----------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------+                      
| Trial name                          | status   | loc                  |   iter |   total time (s) |      ts |   reward |   episode_reward_max |   episode_reward_min |   episode_len_mean |                      
|-------------------------------------+----------+----------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------|                      
| APPO_PongNoFrameskip-v4_639fd_00000 | RUNNING  | 192.168.1.74:4189099 |   1167 |          13934.2 | 3046000 |   -20.22 |                  -17 |                  -21 |            3747.82 |                      
+-------------------------------------+----------+----------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------+                      
                                                                                                         
                                                                                                                                                                                                                   
(RolloutWorker pid=4189081) [2022-02-22 21:32:41,252 E 4189081 4189081] plasma_store_provider.cc:122: Failed to put object bbb0ade83413523bb3b6e75bbebcc7d86b1d79350100000001000000 in object store because it is full. Object size is 11313457 bytes.                                                                                                                                                                                
(RolloutWorker pid=4189081) Plasma store status:                                                                                                                                                                   
(RolloutWorker pid=4189081)                                                                                                                                                                                        
(RolloutWorker pid=4189081) (global lru) capacity: 19458202828                                                                                                                                                     
(RolloutWorker pid=4189081) (global lru) used: 0%                                                                                                                                                                  
(RolloutWorker pid=4189081) (global lru) num objects: 0                                                                                                                                                            
(RolloutWorker pid=4189081) (global lru) num evictions: 42313                                                                                                                                                      
(RolloutWorker pid=4189081) (global lru) bytes evicted: 468750773694                                     
(RolloutWorker pid=4189081) ---                                                                          
(RolloutWorker pid=4189081) --- Tip: Use the `ray memory` command to list active objects in the cluster. 
(RolloutWorker pid=4189081) ---                                                                          
(RolloutWorker pid=4189081)                         
(RolloutWorker pid=4189082) [2022-02-22 21:32:41,252 E 4189082 4189082] plasma_store_provider.cc:122: Failed to put object c504bdc6be5ac365cb0b11874cedf833ae26ec1e0100000001000000 in object store because it is full. Object size is 11313457 bytes.                                                                                                                                                                                
(RolloutWorker pid=4189082) Plasma store status:                                                                                                                                                                   
(RolloutWorker pid=4189082)                                                                                                                                                                                        
(RolloutWorker pid=4189082) (global lru) capacity: 19458202828                                                                                                                                                     
(RolloutWorker pid=4189082) (global lru) used: 0%                                                                                                                                                                  
(RolloutWorker pid=4189082) (global lru) num objects: 0                                                                                                                                                            
(RolloutWorker pid=4189082) (global lru) num evictions: 42313                                                                                                                                                      
(RolloutWorker pid=4189082) (global lru) bytes evicted: 468750773694                                                                                                                                               
(RolloutWorker pid=4189082) ---                     
(RolloutWorker pid=4189082) --- Tip: Use the `ray memory` command to list active objects in the cluster. 
(RolloutWorker pid=4189082) ---                                                                          
(RolloutWorker pid=4189082)                                                                              
```

I can reproduce the same behaviour with Tensorflow and Pytorch.


For all these experiment I used:
```
rllib/tuned_examples/impala/pong-impala-fast.yaml
rllib/tuned_examples/ppo/pong-ppo.yaml
rllib/tuned_examples/ppo/pong-appo.yaml
```

### Versions / Dependencies

Ray: 3cb858 (latest master)
Ubuntu: 20.04
Tensorflow: 2.8.0
Pytorch: 1.10.2

### Reproduction script

```
python rllib/train.py -f <exp file>
```
with `exp file`:
```
rllib/tuned_examples/impala/pong-impala-fast.yaml
rllib/tuned_examples/ppo/pong-ppo.yaml
rllib/tuned_examples/ppo/pong-appo.yaml
```

### Anything else

Every time

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by vakker** [comment]
@sven1977 thanks for the additional details, the configs are the same that are in `tuned_examples` (except the `num_cpus_per_worker: 0.5` bit):
<details>
<summary>ppo/pong-appo.yaml</summary>

```
pong-appo:
    env: PongNoFrameskip-v4
    run: APPO
    stop:
        episode_reward_mean: 18.0
        timesteps_total: 5000000
    config:
        # Works for both torch and tf.
        framework: tf
        vtrace: True
        use_kl_loss: False
        rollout_fragment_length: 50
        train_batch_size: 750
        num_workers: 32
        broadcast_interval: 1
        max_sample_requests_in_flight_per_worker: 1
        num_multi_gpu_tower_stacks: 1
        num_envs_per_worker: 8
        num_cpus_per_worker: 0.5
        minibatch_buffer_size: 4
        num_sgd_iter: 2
        vf_loss_coeff: 1.0
        clip_param: 0.3
        num_gpus: 1
        grad_clip: 10
        model:
          dim: 42
```
</details>

<details>
<summary>impala/pong-impala-fast.yaml</summary>

```
pong-impala-fast:
    env: PongNoFrameskip-v4
    run: IMPALA
    config:
        framework: torch
        rollout_fragment_length: 50
        train_batch_size: 1000
        num_workers: 32
        num_envs_per_worker: 5
        num_cpus_per_worker: 0.5
        broadcast_interval: 5
        max_sample_requests_in_flight_per_worker: 1
        num_multi_gpu_tower_stacks: 4
        num_gpus: 1
        model:
          dim: 42
```
</details>

What config are used for your tests?
Could you possibly provide the timesteps and the wall time it takes to reach those rewards? Or even better the learning curves?
I can narrow it down on my end if this seems to be non-reproducible on your end.

**Comment by vakker** [comment]
@sven1977 could you provide the configs that you use for your tests? Thanks

**Comment by gjoliver** [comment]
we use this config in our weekly test:
https://github.com/ray-project/ray/blob/master/release/rllib_tests/learning_tests/hard_learning_tests.yaml#L91-L116

I'd focus on torch APPO first, since I may have seen it "not learning" as well.
do you see the same problem with torch 1.9?

**Comment by vakker** [comment]
@gjoliver thanks for the additional info, I tried to debug this further. I used the official `rayproject/ray-ml:1.12.0-py38-cu112` docker image to rule out any version or OS issues.

I run the benchmark (APPO Pong) on 2 machines, 1 has a 1080ti the other has a 2080ti. It seems like the performance (reward) correlates with the performance of the system. They were both run for 5M steps, the plots below show the wall time on the x axis (hence the different length). I've run the same training configs (that you linked) 20 times on each machine.

Sample throughput:
![sample_throughput](https://user-images.githubusercontent.com/3268882/162617544-3dc0a555-b3a1-49dc-a7ba-2bd8de2257d5.png)

Reward:
![reward](https://user-images.githubusercontent.com/3268882/162617546-497a0693-b7ed-4c14-a681-ca0ca280e0be.png)

The training does reach the threshold (~18) on the faster machine on occasions (there's still a large variance), but to me it seems like the different sample throughput impacts the training. Can that be the case? #17490 might be related.

I would expect the training to be slower with an older GPU, but the reward curves should be the same (if plotted against the environment samples).

**Comment by gjoliver** [comment]
Interesting. So everything is identical, except for the speed of the GPUs?
What's the x-axis of your episode_reward_mean chart? Is it sample step?
The GPU is used by the trainer right? The workers are doing rollout without GPU I assume. Do you have the same # of workers for both machinese?

**Comment by vakker** [comment]
Everything is *almost* identical. The number of workers used is the same (10), but one is an Intel CPU with 20 cores, the other is an AMD with 24 cores, but the performance difference per CPU should be marginal. The CPU utilization also shows that (approx half for the slower GPU machine).
And yes, GPU is for the trainer only, rollouts are CPU only.

The x-axis is time (in hours), each experiment was run for 5M sample steps, here's the plot with the sample step on the x-axis:
![reward2](https://user-images.githubusercontent.com/3268882/162908759-292e0ebb-5896-4ea4-81e6-774e879832f3.png)

I'm wondering whether the effective samples per policy update is different if the GPU is slower to process the samples.
Maybe it's an interaction of:
```
    # Max queue size for train batches feeding into the learner.
    "learner_queue_size": 16,
    # Wait for train batches to be available in minibatch buffer queue
    # this many seconds. This may need to be increased e.g. when training
    # with a slow environment.
    "learner_queue_timeout": 300,
```
E.g. the samples time out in the queue, so the trainer sees a different distribution of samples compared to what it would see on other machines.

**Comment by gjoliver** [comment]
thanks again for more data.
we found it suspicious that the gpu utilization for APPO is so low.
any chance you can upload the entire tensorboard data to https://tensorboard.dev/, so that we can look at some of the stats on our end??
there is probably something wrong with the learner queue for APPO agent.

**Comment by vakker** [comment]
Sure, here you go: https://tensorboard.dev/experiment/0ArEjoS4TdiAynVa6LHmzQ/
Please note, that I couldn't reproduce the same low GPU utilization that I included in the first post, I'm not sure why that happened. 
In the uploaded Tensorboard logs each experiment was run on GPU-0, so `gpu_util_percent0` is the relevant one, I had something else running on GPU-1, so ignore that.

Each experiment under `appo-pongnoframeskip-v4-other` are with the 2080ti, and `
appo-pongnoframeskip-v4-wombat` are with the 1080ti.

Please let me know if you need anything else, or you want me to rerun any particular experiment.

**Comment by gjoliver** [comment]
thanks this is very helpful

**Comment by vakker** [comment]
@gjoliver Have you found anything that might be causing this issue? Do you need any further information?

**Comment by vakker** [comment]
Do you have any update on this? Probably a lot has changed since this issue was opened.
Have you done any changes to the sample queues? 
I was thinking that "throttling" the rollout workers to match the consumption speed (which can change due to different hardware and network properties) might be a way to reach maximum throughput for a given setup.

E.g.
1. option: GPU is processing the samples slower than the samplers can produce: only sample x batches ahead. This way the CPUs are going to be idle, but then that's a signal to the user that that's a GPU bottleneck. 
2. option: rollout workers are producing samples slower than what the GPU can consume: GPU will be idle, which shows a CPU bottleneck.

Either way, it should lead to maximum resource utilization (taking into account the bottlenecks), it's still async, produces reproducible results regardless of the infrastructure, and it avoids object store fill-up. 

What do you think?

**Comment by gjoliver** [comment]
hey @vakker, thanks for the message. we didn't really find anything last time.
about throttling either GPU or rollout workers so that their throughputs match, personally I had found it to be really important for a successful training run. otherwise you are wasting capacity at either trainer or sampler side for algorithms like IMPALA and APPO.
I think num_workers is a parameter that everyone should tune carefully to achieve max throughput and best learning performance.

Btw, for some of the off-policy algorithms, we also have this parameter training_intensity, that regulates the throughput between trainer and sampling workers.

Also, thanks for your repro script. it helped us reproduce the mem leak locally, and was super helpful.
